### DISCUSSION TITLE
Interpretation of case-specific variable importance

### DISCUSSION THREAD

**Discussion by qingnanl** [kogalur/randomForestSRC#268] [Q&A] [answered]
Hello rfsrc developers:

Thanks for making the comprehensive and powerful tool. I am very interested in the function of the case-specific VIMP implemented in this tool. I have a question regarding this function and wish to get some insights from you:

I would like to understand how the CSV is achieved. For example, if I am interested in the CSV for 'Sample1' among Sample1-Sample50, I may find the VIMPs in the original forest, and also the VIMPs in a subset of the original forest, which Sample1 is excluded. Then some types of subtraction could be made between the first set of VIMPs and the second to learn the contribution of 'Sample1'.  Is that what you have done? Or do you have a formula somewhere that I could learn about it?

Many thanks,

qingnanl

**Comment by ishwaran** [comment, marked as answer]
VIMP is obtained by averaging over cases and trees.  Consider a single tree.  About .368 (~37%) of the cases are out-of-bag (OOB) which means these cases were not used for constructing the tree and therefore they can be used for determining prediction error and VIMP.  The VIMP for these 37% cases are calculated and these are then averaged to yield the tree VIMP.   Averaging the tree VIMP over all trees results in the final VIMP value.

Now what about a specific case, say case #1?  Over 1000 trees we can imagine there will 370 trees where case #1 is OOB.  We calculate the VIMP for case #1 for each of its 370 OOB trees and then average these values.  The averaged value is what is called the case specific VIMP for case #1.

**Reply by qingnanl** [reply]
Thanks for the reply. What I am thinking is to infer the 'contribution' of each sample (being learned) toward the RF model, so maybe I should do some 'subtractions' for the VIMP in trees with (1000) and without (370) a certain case. Does that make any sense to you?

Thanks,
qingnanl

**Reply by ishwaran** [reply]
Then it seems you are not interested in VIMP (which means variable importance).  It seems you are interested in the contribution of a case to the overall performance of the model.  

The following example shows how the overall error rate of the forest can be obtained directly from the OOB predicted value:

First call the forest and output the results
```
> o <- rfsrc(mpg~., mtcars)
> print(o)

                         Sample size: 32
                     Number of trees: 500
           Forest terminal node size: 5
       Average no. of terminal nodes: 3.462
No. of variables tried at each split: 4
              Total no. of variables: 10
       Resampling used to grow trees: swor
    Resample size used to grow trees: 20
                            Analysis: RF-R
                              Family: regr
                      Splitting rule: mse *random*
       Number of random split points: 10
                     (OOB) R squared: 0.77246194
   (OOB) Requested performance error: 8.26511574

```

The prediction error rate can also be obtained as follows

```
> print(get.mv.error(o))
     mpg 
8.265116 
```

Here we do a direct calculation to get the error rate

```
> mean((o$yvar-o$predicted.oob)^2)
[1] 8.265116
```

Therefore we can see that the OOB predicted value `o$predicted.oob` contains the (out-of-sample) contribution of each case to the prediction of the forest.  I think this is what you are asking about.

**Reply by qingnanl** [reply]
Thanks for the quick response! Sorry I haven’t made it clear. I am interested in the VIMP, but particularly to ‘how each sample contributed to the VIMP of each feature’. That’s why I was talking about the ‘subtraction’ thing.

qingnanl

**Reply by ishwaran** [reply]
Then the answer is provided above.  From the help file, this is how it is obtained:
 
```
## ------------------------------------------------------------
## case-specific vimp
## returns VIMP for each case
## ------------------------------------------------------------
     
o <- rfsrc(mpg~., mtcars)
op <- predict(o, importance = TRUE, csv = TRUE)
csvimp <- get.mv.csvimp(op, standardize=TRUE)
print(csvimp)

```

**Reply by qingnanl** [reply]
Thanks. The reason I am keeping bringing up the subtraction is that, the current csv options is calculating the VIMP using the trees where Case1 is OOB. Thus for these trees, Case1 is among the OOB samples to evaluate the performance of them (together with other OOB samples). However, to my understanding, Case1 is not used in the learning process when we grow these trees (370 if 1000 in total). Thus, regarding ‘how each sample contributed to the VIMP of each feature’, do you think it makes some sense to compare the VIMP of trees with and without a certain case (e.g. Case1) during the training process (growing the tree)? Thanks again for your kind help and your ideas.

qingnanl

**Reply by ishwaran** [reply]
No, what you are proposing won't work because it is exactly when the case is OOB that we can use that tree for the calculation of VIMP.  The thing is that VIMP (classical Breiman-Cutler) doesn't measure the contribution of a case,  or even a variable, to the learning process at all.  That would mean that we are somehow measuring the importance of a variable in the RF predictor but actually we don't do that.   VIMP is a tree based calculation.  It uses a single tree to determine the importance of a variable and then it averages these values (the latter being done to reduce variance).  

That is to say, _**VIMP is a variable selection technique that is actually unrelated to the RF ensemble**_.  It is a method for identifying variables that are informative for the true underlying functional of interest (which we don't know).

That being said, CSV is very useful for understanding how much a case is related to the true functional of interest.  

I think for you to understand this, you need to separate the idea of understanding "the learning process" from the actual goal of VIMP, which is to understand the true functional.

**Reply by qingnanl** [reply]
Thanks, and sorry, I was quite distracted today. I was thinking about using VIMP as a metric to measure the association between features (one as a predictor and one as a target, of course). Along this line, the 'case-specific VIMP' I am trying to obtain could be an interesting measurement of the association between features in local cases. I guess it is more or less how the VIMP in a model with a case compares with one excluding it. However, training such models could be computationally expensive in large datasets (need to train a model for each sample (to exclude it)). Then I started to think maybe it is possible to play with the tree-wise VIMP and compare those with and without a sample due to the subsampling. In this way, the training only needs to be done once and all the information is there. I should be able to do some testing.  

qingnanl
